File: ICPC 2018-19/Swamp_County_Toll_Roads.py
'''
SWAMP COUNTY TOLL ROADS
Concepts: dictionary, decimal precision/rounding
Author: David Van Heck

Start Time: 2:50
End Time: 3:15

'''

import sys

prob_dict = dict() # stores the letter as keys and the probabilities as values

# read probablilities and store them in dictionary
line = sys.stdin.readline()
while line.strip() != '':
    letter_and_prob = line.replace('\n', '').split(' ')
    prob_dict[letter_and_prob[0]] = float(letter_and_prob[1])
    line = sys.stdin.readline()

# read license plates and output probabilites
for license_plate in sys.stdin:
    prob = 1 # probability of license plate being correct
    for letter in license_plate:
        if letter in prob_dict:
            prob *= prob_dict[letter]
    print('%.3f'%round(prob, 3))

File: ICPC 2018-19/SwampCountyTollRoads.cpp
/* 
   INEFFICIENT SWAMP COUNTY TOLL ROADS WITHOUT USING MAP/DICTIONARY
   Concepts: decimal precision, vectors
   Author: David Van Heck
*/ 
#include <iostream>
#include <string>
#include <vector>
#include <iomanip>

using namespace std;

int main()
{
  string input; // stores input
  vector<char> letters; // stores license plate letters
  vector<float> probs; // stores probabilities
  
  // store letters and probablilites
  getline(cin, input);
  for(int i = 0; !input.empty(); i++)
  {
      letters.push_back(input.substr(0, 1).at(0));
      probs.push_back(stof(input.substr(2, input.size())));
      
      getline(cin, input);
  }
  
  // calculate probability of each license plate
  getline(cin, input);
  while(!cin.eof())
  {
    float prob = 1.0;

    // iterate through each letter in the license plate
    for(int i = 0; i < input.size(); i++) 
    {
        // check if the current letter has a corresponding probability, if it does multiply by the probability
        bool letterFound = false;
        for(int j = 0; j < letters.size() && letterFound == false; j++)
        {
            if(input[i] == letters[j])
            {
                prob *= probs[j];
                letterFound == true;
            }
        }
    }
    
    // output probability
    cout << setprecision(3) << fixed << prob << endl;
    
    // read next license plate
    getline(cin, input);
  }
}